consensus/istanbul/proxy/types.go

```diff
@@ -0,0 +1,165 @@
+
+	// errInvalidEnodeCertificate is returned if the enode certificate is invalid
+	errInvalidEnodeCertificate = errors.New("invalid enode certificate")
+
+	// errUnauthorizedProxiedValidator is returned if the peer connecting is not the
+	// authorized proxied validator
+	errUnauthorizedProxiedValidator = errors.New("unauthorized proxied validator")
+
+	// ErrStoppedProxyHandler is returned if proxy handler is stopped
+	ErrStoppedProxyHandler = errors.New("stopped proxy handler")
+
+	// ErrStartedProxyHandler is returned if proxy handler is already started
+	ErrStartedProxyHandler = errors.New("started proxy handler")
+
+	// ErrNodeNotProxiedValidator is returned if this node is not a proxied validator
+	ErrNodeNotProxiedValidator = errors.New("node not a proxied validator")
+
+	// ErrNodeNotProxy is returned if this node is not a proxy
+	ErrNodeNotProxy = errors.New("node not a proxy")
+)
+
+type ProxyEngine interface {
+	Start() error
+	Stop() error
+	HandleMsg(peer consensus.Peer, msgCode uint64, payload []byte) (bool, error)
+	AddProxy(node, externalNode *enode.Node) error
+	RemoveProxy(node *enode.Node) error
+	RegisterProxyPeer(proxyPeer consensus.Peer) error
+	UnregisterProxyPeer(proxyPeer consensus.Peer)
+	RegisterProxiedValidatorPeer(proxiedValidatorPeer consensus.Peer)
+	UnregisterProxiedValidatorPeer(proxiedValidatorPeer consensus.Peer)
+	SendEnodeCertificateMsgToProxiedValidator(msg *istanbul.Message) error
+	SendForwardMsg(finalDestAddresses []common.Address, ethMsgCode uint64, payload []byte, proxySpecificPayload map[enode.ID][]byte) error
+	// SendDelegateSignMsgToProxy(msg []byte) error
+	// SendDelegateSignMsgToProxiedValidator(msg []byte) error
+	SendValEnodesShareMsg(proxyPeer consensus.Peer, remoteValidators []common.Address) error
+	SendValEnodesShareMsgToAllProxies()
+	GetValidatorProxyAssignments() (map[common.Address]*enode.Node, error)
+}
+
+// ==============================================
+//
+// define the proxy object
+
+type proxy struct {
+	node         *enode.Node    // Enode for the internal network interface
+	externalNode *enode.Node    // Enode for the external network interface
+	peer         consensus.Peer // Connected proxy peer.  Is nil if this node is not connected to the proxy
+	disconnectTS time.Time      // Timestamp when this proxy's peer last disconnected. Initially set to the timestamp of when the proxy was added
+}
+
+// ProxyInfo is used to provide info on a proxy that can be given via an RPC
+type ProxyInfo struct {
+	InternalNode *enode.Node      `json:"internalEnodeUrl"`
+	ExternalNode *enode.Node      `json:"externalEnodeUrl"`
+	IsPeered     bool             `json:"isPeered"`
+	Validators   []common.Address `json:"validators"`            // All validator addresses assigned to the proxy
+	DisconnectTS int64            `json:"disconnectedTimestamp"` // Unix time of the last disconnect of the peer
+}
+
+func (p proxy) ID() enode.ID {
+	return p.node.ID()
+}
+
+func (p proxy) Info() ProxyInfo {
```

Comment:
nit: I would invert the relationship, an have:
```go
newProxyInfo(proxy, validators)
```
with or without the validators.

The reason is that in proxy_set you call `Info()` and then mutate the result, which looks weird, since info seems to be a 
"view" of the proxy which you shouldn't be mutating (since you don't own it)

Reply:
Yep.  Makes sense.